include/mxnet/io.h

```diff
@@ -113,5 +120,105 @@ struct DataIteratorReg
  */
 #define MXNET_REGISTER_IO_ITER(name)                                    \
   DMLC_REGISTRY_REGISTER(::mxnet::DataIteratorReg, DataIteratorReg, name)
+
+/*!
+ * \brief A random accessable dataset which provides GetLen() and GetItem().
+ * Unlike DataIter, it's a static lookup storage which is friendly to random access.
+ * The dataset itself should NOT contain data processing, which should be applied during
+ * data augmentation or transformation processes.
+ */
+class Dataset {
+  public:
+    /*!
+    *  \brief Initialize the Operator by setting the parameters
+    *  This function need to be called before all other functions.
+    *  \param kwargs the keyword arguments parameters
+    */
+    virtual void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) = 0;
```

Comment:
As "This function need to be called before all other functions", why not use a constructor instead? That way compiler will ensure initialization happens.

Reply:
Due to the argument parsing system mxnet uses, python has to create object first and pass arguments to call `Init`

Reply:
`MXDatasetCreateDataset` writes the address of the created dataset to the `out` argument. Why is there be a requirement to have an `Init` function?

Reply:
This is still outstanding. Do you have a plan to fix the API design issue?